### PR TITLE
BUG: remove fast scalar power for arrays with object dtype

### LINKED ISSUE #11047: BUG: ndarray.__pow__ dispatches to square also for object dtype
`ndarray.__pow__` dispatches to shortcuts for simple powers such as `-1, 0, 0.5, 1, and 2`. As noted by @numpynewb (https://github.com/numpy/numpy/issues/9363#issuecomment-386657349), this is not reasonable for an array with object dtype, and for most exponents, the short-cuts are not used, except for exponent = 2, where `square` is called instead of `pow(.., 2)` for all `dtype`. See  https://github.com/numpy/numpy/blob/b9e3ac9abb6e435cdf6bbe33e0bc894d6a879a53/numpy/core/src/multiarray/number.c#L505

### PR DESCRIPTION
**Pull request by nlee737**
Fixes #11047.

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
Looks good, but please do add a test!

**Comment by numpynewb** [comment]
This is maybe not the place to ask, so apologies if that is the case, but is there a place where I can learn about how to add a test for this specific patch? I would love to contribute effort on my end to help move this along, if a test is holding up a merge. Thanks, and again apologies if this not the place; I can delete this message if so.

**Comment by mattip** [comment]
Since you are touching code that handles `power`, you could add it next to `test_pow_override_with_errors` in `test_multiarray.py`. You can add a new class method that starts with `test_`, it should have an `assert_` or `assert_equal` or so to show something that failed before, and now passes.

**Comment by fivemok** [comment]
Thank you for the guidance. I'll add the tests shortly when I get a chance.

**Comment by mhvk** [comment]
@numpynewb, @fivemok - this is absolutely the place to ask! And @mattip, thanks for stepping in and answering.

**Comment by numpynewb** [comment]
Thank you! And, it looks like @fivemok beat me to it.

I am not sure I understand the following assertion yet:

```python
        with assert_raises(TypeError):
            np.square(obj_arr)
```

I think that the `TypeError` will be raised in both the patched and unpatched versions, because the power function is never invoked when calling `np.square`. As far as I can tell, `np.square` looks like an [alias for self-multiplication](https://github.com/numpy/numpy/blob/d407f24cb5ad85850a60b1be3dade3305ea30c98/numpy/core/src/umath/funcs.inc.src#L21), but I could be totally wrong about that.

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -3332,7 +3332,43 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kw):
         with assert_raises(NotImplementedError):
             a ** 2
 
-
+    def test_fast_scalar_pow(self):
+        # test for fast exponentiation on arrays
+        class SomeClass(object):
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: A comment explaining that `__mul__` and `__div__` are deliberately missing, and that this tests that `**` is _not_ taking a fast path via them.

Better would be to add`
```
def __mul__(self, other):
    raise AssertionError('__mul__ should not be called')
```

**Comment by fivemok** [comment]
@eric-wieser Thanks for the feedback. I'll make the appropriate changes shortly.

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -3332,7 +3332,43 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kw):
...
+
+        assert_equal(num_arr ** 0.5, pow_for(0.5, num_arr))
+        assert_equal(num_arr ** 0, pow_for(0, num_arr))
+        assert_equal(num_arr ** 1, pow_for(1, num_arr))
+        assert_equal(num_arr ** -1, pow_for(-1, num_arr))
+        assert_equal(num_arr ** 2, pow_for(2, num_arr))
+
+        # fast pow operations should only be invoked on arrays of numeric dtype
+        with assert_raises(TypeError):
+            np.square(obj_arr)
```
[the remark this answers is not in the document]
Comment (fivemok) [reply numpy/core/tests/test_multiarray.py]: Should I only test on arrays of object dtype?

[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: I meant just this `np.square` call being not that relevant.

Regarding testing only object arrays - I think that you're right with that, and there's not much point testing numeric arrays here - they're already tested elsewhere, I imagine.

**Comment by fivemok** [comment]
@numpynewb I consulted the [Python docs](https://docs.python.org/3/c-api/number.html#c.PyNumber_Multiply) and I think you are right. Thanks for pointing this out.

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Thanks @fivemok1

**Comment by numpynewb** [comment]
Thank you @fivemok!

**Pull request merged**